service/service.go

```diff
@@ -170,59 +173,96 @@ func NewService(cfg *Config, datadir string) (*Service, error) {
 		minMemoryLayer:  uint(minMemoryLayer),
 		genesis:         genesis,
 		datadir:         datadir,
-		executingRounds: make(map[string]*round),
-		errChan:         make(chan error, 10),
+		executingRounds: make(map[string]struct{}),
 		privKey:         privateKey,
 		PubKey:          privateKey.Public().(ed25519.PublicKey),
 	}
+
 	log.Info("Service public key: %x", s.PubKey)
 
 	return s, nil
 }
 
-func (s *Service) loop(ctx context.Context) {
-	var executingRounds errgroup.Group
-	defer executingRounds.Wait()
+type roundResult struct {
+	round *round
+	err   error
+}
+
+func (s *Service) loop(ctx context.Context, roundsToResume []*round, cmds <-chan Command) {
+	var eg errgroup.Group
+	defer eg.Wait()
+
+	logger := log.AppLog.WithName("worker")
+	ctx = logging.NewContext(ctx, logger)
+
+	roundResults := make(chan roundResult, 1)
+
+	// Resume recovered rounds
+	for _, round := range roundsToResume {
+		round := round
+		s.executingRounds[round.ID] = struct{}{}
+		end := s.roundEndTime(round)
+		eg.Go(func() error {
+			err := round.recoverExecution(ctx, round.stateCache.Execution, end)
+			roundResults <- roundResult{round: round, err: err}
+			return nil
+		})
+	}
+
+	timer := s.scheduleRound(ctx, s.openRound)
 
 	for {
-		s.openRoundMutex.RLock()
-		epoch := s.openRound.Epoch()
-		s.openRoundMutex.RUnlock()
-
-		start := s.genesis.Add(s.cfg.EpochDuration * time.Duration(epoch)).Add(s.cfg.PhaseShift)
-		waitTime := time.Until(start)
-		timer := time.After(waitTime)
-		if waitTime > 0 {
-			log.Info("Round %v waiting for execution to start for %v", s.openRoundID(), waitTime)
-		}
 		select {
+		case cmd := <-cmds:
+			cmd(s)
```

Comment:
i would do eg.Go() to execute commands. we don't know if the passed in func can block for a long time or not.